Refactor from_arrow_device/host to use resource_ref

harrism opened the pull request:
## Description
Fixes #16159

Also fixes typos / leftovers in  dictionary `add_keys` copydocs.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

vyasr reviewed (commented):
My apologies, #15904 (which was based on the PRs adding the code that this fixed) proliferated the issue (and introduced the conflicts here). When resolving conflicts, could you fix the `from_arrow_stream` API as well? I'll be careful to correctly handle the mrs in future PRs.

vyasr commented on cpp/include/cudf/interop.hpp:
```diff
@@ -391,8 +392,8 @@ std::unique_ptr<cudf::table> from_arrow(
   ArrowArray const* input,
-  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
-  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+  rmm::cuda_stream_view stream,
```
We lost the default here (I'd add a suggestion but I'll probably screw up the spacing in the GUI).

harrism replied:
Good catch, thanks. Fixed.

harrism commented:
Can't see why the docs build is failing. Warnings only: 

```
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/dataframe.py:docstring of cudf.core.dataframe.DataFrame.to_csv:35: WARNING: py:data reference target not found: os.linesep

/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/indexed_frame.py:docstring of cudf.core.indexed_frame.IndexedFrame.to_hdf:71: WARNING: py:func reference target not found: open
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/indexed_frame.py:docstring of cudf.core.indexed_frame.IndexedFrame.to_hdf:71: WARNING: py:func reference target not found: open
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/column/string.py:docstring of cudf.core.column.string.StringMethods.istitle:6: WARNING: py:meth reference target not found: str.istitle
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/column/string.py:docstring of cudf.core.column.string.StringMethods.join:5: WARNING: py:meth reference target not found: str.join
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/column/string.py:docstring of cudf.core.column.string.StringMethods.rsplit:49: WARNING: py:obj reference target not found: str.split
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/column/string.py:docstring of cudf.core.column.string.StringMethods.rsplit:51: WARNING: py:obj reference target not found: str.rsplit
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/column/string.py:docstring of cudf.core.column.string.StringMethods.split:50: WARNING: py:obj reference target not found: str.split
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/core/column/string.py:docstring of cudf.core.column.string.StringMethods.split:52: WARNING: py:obj reference target not found: str.rsplit
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/io/hdf.py:docstring of cudf.io.hdf.read_hdf:10: WARNING: py:class reference target not found: pathlib.Path
/opt/conda/envs/docs/lib/python3.11/site-packages/cudf/io/hdf.py:docstring of cudf.io.hdf.read_hdf:46: WARNING: py:func reference target not found: open
```

harrism commented:
/merge
